### DISCUSSION TITLE
M5stack ESP32 rs485 adaptor?

### DISCUSSION THREAD

**Discussion by ywteh** [nathanmarlor/foxess_modbus#654] [Q&A]
Hi, apologies for newbie question, but I'm wondering if this integration could be made to work using an m5stack Atom RS485 to TTL/Wifi converter like this: https://shop.m5stack.com/products/atom-rs485-kit?variant=34787900194980? If so, what would be required to make it work? Thank you!

**Comment by FozzieUK** [comment]
No that board is designed to be interfaced with a dev board or ESPHome device where it interfaces to directly using TTL, this integration has been designed to use direct ethernet, TCP Modbus or USB modbus devices which interface easily with home assistant. 

Nothing is impossible but the amount of work to interface it just wouldn't be credible.

**Comment by canton7** [comment]
It looks like that's an ESP32 microprocessor, connected to an RS-485 level converter?

If you fancy writing some software which runs on the ESP32 which talks USB and pretends to be a serial device, I'm sure you could make it work. There are probably some Arduino libraries which make it fairly easy. At that point, you could just connect as with any other USB to serial adapter.

But, given that you can buy something which already does this for a few dollars, is that really worth the time?

**Comment by rand12345** [comment]
@ywteh

There's a few options here:

1. You can flash the Atom with [Tasmota Zbbridge firmware](https://tasmota.github.io/install/) using a Google Chrome web browser and then configure the device as a serial to TCP bridge as described [here.](https://tasmota.github.io/docs/Serial-to-TCP-Bridge/) This add-on is then configured as ```Ethernet to Modbus/RS485 Adapter``` -> ```Other``` -> ```Modbus RTU over TCP``` 

    The Atom seems to use a 4 wire TTL interface which means you may need to assign a [TCP TX enable pin](https://github.com/arendst/Tasmota/pull/21269) too. However, in the past I have simply pulled the DE pin high in the Tasmota GPIO settings and all seemed ok.

2. Alternatively, if you want to use Tasmota as an Modbus TCP to RTU adaptor then you will need to recompile the Tasmota firmware (a headache).

    ```USE_MODBUS_BRIDGE_TCP``` would need to be enabled as discussed [here.](https://tasmota.github.io/docs/Modbus-Bridge/)

3. If you don't have much luck with Tasmota (the best solution I think) then create a simple Arduino sketch which simply pipes data from a TCP socket to Serial1 and vice-versa. This add-on should then be configured as the first option above. Example code below should be close to what you need.

``` c

#include <WiFi.h>

#define RX_PIN 4
#define TX_PIN 6
#define DE_PIN 2  // Define the DE pin for RS485
#define RE_PIN 3  // Define the RE pin for RS485
#define BAUDRATE 9600

// wifi config
const char* ssid = "SSID";
const char* password = "PASSWORD";

// ethernet config
const IPAddress local_IP(10, 0, 10, 249);
const IPAddress gateway(10, 0, 10, 1);
const IPAddress subnet(255, 255, 255, 0);
const IPAddress primaryDNS(8, 8, 8, 8);
const IPAddress secondaryDNS(8, 8, 4, 4);

const int serverPort = 502;

#define BUFFER_SIZE 1024

// global objects
WiFiServer server;
byte buff[BUFFER_SIZE];

void setup() {
  pinMode(DE_PIN, OUTPUT);
  pinMode(RE_PIN, OUTPUT);
  digitalWrite(DE_PIN, LOW);  // Default to receive mode
  digitalWrite(RE_PIN, LOW);  // Default to receive mode
  // init serial ports
  Serial.begin(115200);
  Serial1.begin(BAUDRATE, SERIAL_8N1, RX_PIN, TX_PIN);

  // init wifi connection
  if (!WiFi.config(local_IP, gateway, subnet, primaryDNS, secondaryDNS)) {
    Serial.println("Failed to configure network settings");
  }
  WiFi.begin(ssid, password);
  while (WiFi.status() != WL_CONNECTED) {
    Serial.printf("connecting to WiFi network %s\n\r", ssid);
    delay(500);
  }

  Serial.println("connected to WiFi");
  Serial.print("IP adddr: ");
  Serial.println(WiFi.localIP());
  delay(1000);

  //start server
  server = WiFiServer(serverPort);
  server.begin();
  delay(1000);
  Serial.println("server started");
}

void loop() {
  // wait for client
  WiFiClient client = server.available();
  if (!client)
    return;

  Serial.println("client connected");
  while (client.connected()) {
    int size = 0;

    // read data from wifi client and send to serial
    while ((size = client.available())) {
      size = (size >= BUFFER_SIZE ? BUFFER_SIZE : size);
      client.read(buff, size);
      Serial.printf("TCP->SER %d ", size);
      for (int i = 0; i < size; i++) {
        Serial.printf("0x%02x, ", buff[i]);
      }
      Serial.println();
      // Enable RS485 driver (DE high, RE low) to transmit
      digitalWrite(DE_PIN, HIGH);
      Serial1.write(buff, size);
      Serial1.flush();
      // Disable RS485 driver (DE low, RE low) after transmission
      digitalWrite(DE_PIN, LOW);
    }

    // read data from serial and send to wifi client
    while ((size = Serial1.available())) {
      size = (size >= BUFFER_SIZE ? BUFFER_SIZE : size);
      Serial1.readBytes(buff, size);
      Serial.printf("TCP<-SER %d ", size);
      for (int i = 0; i < size; i++) {
        Serial.printf("0x%02x, ", buff[i]);
      }
      Serial.println();
      client.write(buff, size);
      client.flush();
    }
  }
  Serial.println("client disconnected");
  client.stop();
}


```

**Comment by pantherale0** [comment]
Sorry to dig up an old discussion... Add these lines to an esphome config file and it should work:

```yaml
external_components:
  - source: github://wolffshots/esphome-stream-server

# Specify pins used by the board to comunicate with RS485 board - in this case M5STACK
uart:
  tx_pin: GPIO19
  rx_pin: GPIO22
  baud_rate: 9600
  parity: NONE
  data_bits: 8
  stop_bits: 1
  id: uart_bus

stream_server:
  buffer_size: 512
  uart_id: uart_bus
  port: 502
```
